qiskit/backends/ibmq/ibmqprovider.py

```diff
@@ -26,6 +28,16 @@ def __init__(self, token, url,
         # Populate the list of remote backends.
         self.backends = self._discover_remote_backends()
 
+        # Add name to Provider instance
+        if provider_name is None:
+            if 'quantumexperience' in url:
+                provider_name = 'ibmq'
+            elif 'q-console' in url:
+                provider_name = 'qnet'
+            else:
+                raise QISKitError('Unkown IBMQ provider.')
```

Comment:
Following up on the comment in `BaseProvider`, I think this introduces some tiny yet undesirable side effects: in particular, makes `provider_name` to be effectively required when creating an `IBMQProvider` instance if using non-standard URLs without real benefits when using the provider directly - ie. the `.name` is only relevant for `register()` and their related functions, which compromises a bit the decoupling. It also seems that in this revision, it is possible for two instances to have the same `name`, which I'm not sure if it is by design and would require some clarification (ie. what is the relation between a provider and its name - always the same name, as happens with the `LocalProvider`, a default name but potentially dependent on the user needs, as in this one, etc?).

Instead of that - what about delegating the handling of provider names to `DefaultQISKitProvider`? It is arguably the purpose of that class, and it seems that by using a "dict of provider_name:  Provider" there would be enough to cover the needs introduced by this PR and at the same time avoid introducing application-dependent changes in the base classes.

Reply:
The name is indeed used only to register and unregister at the moment.  However, as discussed in our meeting, there could be cases where there are multiple `IBMQProvider` instances registered, and I need a way to address them, e.g. to unregister/register one and not the other.  I demoed one such example of this.

If you can think of a better way to accomplish the same task, then I am curious to hear your suggestion.